File: web/src/lib/api/agentDashboard.ts
// API service for Agent Dashboard

export interface TimeRange {
  start_time: string
  end_time: string
}

export interface ProviderModelUsage {
  provider: string
  model: string
  request_count: number
  token_count: number
  cost_usd: number
}

export interface TokenBreakdown {
  input_tokens: number
  output_tokens: number
  thinking_tokens: number
  cost_usd: number
}

export interface TokenMetrics {
  total_input: number
  total_output: number
  total_thinking: number
  total_cost_usd: number
  by_model: Record<string, TokenBreakdown>
}

export interface CacheStatistics {
  total_hits: number
  total_misses: number
  hit_rate: number
  tokens_saved: number
  cost_saved_usd: number
}

export interface ToolCallStatistic {
  tool_name: string
  success_count: number
  failure_count: number
  total_count: number
  success_rate: number
}

export interface PolicyViolationStat {
  violation_type: string
  count: number
  avg_severity: number
}

export interface RiskAssessment {
  score: number
  level: string
  total_violations: number
  details: Record<string, number>
}

export interface AgentDashboardStats {
  api_key_id: string
  api_key_name: string
  time_range: TimeRange
  provider_usage: ProviderModelUsage[]
  token_metrics: TokenMetrics
  cache_stats: CacheStatistics
  tool_call_stats: ToolCallStatistic[]
  violations: PolicyViolationStat[]
  risk_score: RiskAssessment
}

export interface AgentInfo {
  id: string
  name: string
  key_prefix: string
  role_id?: string
  role_name?: string
  group_id?: string
  group_name?: string
  created_at: string
  last_used_at?: string
  revoked: boolean
}

export interface AgentListResponse {
  agents: AgentInfo[]
  total: number
}

// Get auth token from localStorage
function getAuthToken(): string | null {
  return localStorage.getItem('authToken')
}

// Get tenant slug from localStorage
function getTenantSlug(): string | null {
  return localStorage.getItem('tenantSlug')
}

// Base API URL (adjust as needed)
const API_BASE_URL = '/v1'

// Fetch agent dashboard stats
export async function fetchAgentDashboardStats(
  apiKeyId: string,
  startTime: string,
  endTime: string
): Promise<AgentDashboardStats> {
  const token = getAuthToken()
  const tenantSlug = getTenantSlug()
  const params = new URLSearchParams({
    api_key_id: apiKeyId,
    start_time: startTime,
    end_time: endTime,
  })

  const response = await fetch(`${API_BASE_URL}/agents/dashboard/stats?${params}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
      ...(tenantSlug && { 'X-Tenant': tenantSlug }),
    },
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Failed to fetch dashboard stats: ${error}`)
  }

  return response.json()
}

// Fetch risk assessment only
export async function fetchRiskAssessment(
  apiKeyId: string,
  startTime: string,
  endTime: string
): Promise<RiskAssessment> {
  const token = getAuthToken()
  const tenantSlug = getTenantSlug()
  const params = new URLSearchParams({
    api_key_id: apiKeyId,
    start_time: startTime,
    end_time: endTime,
  })

  const response = await fetch(`${API_BASE_URL}/agents/dashboard/risk?${params}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
      ...(tenantSlug && { 'X-Tenant': tenantSlug }),
    },
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Failed to fetch risk assessment: ${error}`)
  }

  return response.json()
}

// List all agents (API keys)
export async function fetchAgentList(): Promise<AgentListResponse> {
  const token = getAuthToken()
  const tenantSlug = getTenantSlug()

  console.log('[Agent Dashboard] Fetching agent list', { token: token ? 'present' : 'missing', tenantSlug })

  const response = await fetch(`${API_BASE_URL}/agents/list`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
      ...(tenantSlug && { 'X-Tenant': tenantSlug }),
    },
  })

  console.log('[Agent Dashboard] Agent list response', { status: response.status, ok: response.ok })

  if (!response.ok) {
    const error = await response.text()
    console.error('[Agent Dashboard] Agent list error', error)
    throw new Error(`Failed to fetch agent list: ${error}`)
  }

  const data = await response.json()
  console.log('[Agent Dashboard] Agent list data', data)
  return data
}

// Helper function to format date to ISO8601
export function formatDateToISO(date: Date): string {
  return date.toISOString()
}

// Helper to get default time range (last 24 hours)
export function getDefaultTimeRange(): { start_time: string; end_time: string } {
  const endTime = new Date()
  const startTime = new Date(endTime.getTime() - 24 * 60 * 60 * 1000)

  return {
    start_time: formatDateToISO(startTime),
    end_time: formatDateToISO(endTime),
  }
}

// Helper to get time range for a specific period
export function getTimeRange(period: '24h' | '7d' | '30d' | 'custom', customStart?: Date, customEnd?: Date): { start_time: string; end_time: string } {
  const endTime = new Date()
  let startTime: Date

  switch (period) {
    case '24h':
      startTime = new Date(endTime.getTime() - 24 * 60 * 60 * 1000)
      break
    case '7d':
      startTime = new Date(endTime.getTime() - 7 * 24 * 60 * 60 * 1000)
      break
    case '30d':
      startTime = new Date(endTime.getTime() - 30 * 24 * 60 * 60 * 1000)
      break
    case 'custom':
      if (!customStart || !customEnd) {
        throw new Error('Custom period requires start and end dates')
      }
      return {
        start_time: formatDateToISO(customStart),
        end_time: formatDateToISO(customEnd),
      }
    default:
      startTime = new Date(endTime.getTime() - 24 * 60 * 60 * 1000)
  }

  return {
    start_time: formatDateToISO(startTime),
    end_time: formatDateToISO(endTime),
  }
}

File: internal/domain/agent_dashboard.go
package domain

import (
	"context"
	"time"
)

// =============================================================================
// Agent Dashboard Types
// =============================================================================

// AgentDashboardStats contains comprehensive agent statistics
type AgentDashboardStats struct {
	APIKeyID   string    `json:"api_key_id"`
	APIKeyName string    `json:"api_key_name"`
	TimeRange  TimeRange `json:"time_range"`

	// Provider and Model Usage
	ProviderUsage []ProviderModelUsage `json:"provider_usage"`

	// Token Metrics
	TokenMetrics TokenMetrics `json:"token_metrics"`

	// Cache Usage
	CacheStats CacheStatistics `json:"cache_stats"`

	// Tool Calls
	ToolCallStats []ToolCallStatistic `json:"tool_call_stats"`

	// Policy Violations
	Violations []PolicyViolationStat `json:"violations"`

	// Risk Assessment
	RiskScore RiskAssessment `json:"risk_score"`
}

// TimeRange represents a time range for querying
type TimeRange struct {
	StartTime time.Time `json:"start_time"`
	EndTime   time.Time `json:"end_time"`
}

// ProviderModelUsage contains usage statistics for a provider/model combination
type ProviderModelUsage struct {
	Provider     string  `json:"provider"`
	Model        string  `json:"model"`
	RequestCount int64   `json:"request_count"`
	TokenCount   int64   `json:"token_count"`
	CostUSD      float64 `json:"cost_usd"`
}

// TokenMetrics contains token usage metrics
type TokenMetrics struct {
	TotalInput    int64                     `json:"total_input"`
	TotalOutput   int64                     `json:"total_output"`
	TotalThinking int64                     `json:"total_thinking"`
	TotalCost     float64                   `json:"total_cost_usd"`
	ByModel       map[string]TokenBreakdown `json:"by_model"`
}

// TokenBreakdown contains token breakdown for a specific model
type TokenBreakdown struct {
	InputTokens    int64   `json:"input_tokens"`
	OutputTokens   int64   `json:"output_tokens"`
	ThinkingTokens int64   `json:"thinking_tokens"`
	CostUSD        float64 `json:"cost_usd"`
}

// CacheStatistics contains cache usage statistics
type CacheStatistics struct {
	TotalHits    int64   `json:"total_hits"`
	TotalMisses  int64   `json:"total_misses"`
	HitRate      float64 `json:"hit_rate"` // Percentage (0-100)
	TokensSaved  int64   `json:"tokens_saved"`
	CostSavedUSD float64 `json:"cost_saved_usd"`
}

// ToolCallStatistic contains statistics for a specific tool
type ToolCallStatistic struct {
	ToolName     string  `json:"tool_name"`
	SuccessCount int64   `json:"success_count"`
	FailureCount int64   `json:"failure_count"`
	TotalCount   int64   `json:"total_count"`
	SuccessRate  float64 `json:"success_rate"` // Percentage (0-100)
}

// PolicyViolationStat contains statistics for policy violations by type
type PolicyViolationStat struct {
	ViolationType string  `json:"violation_type"`
	Count         int64   `json:"count"`
	AvgSeverity   float64 `json:"avg_severity"` // Average severity (1-5)
}

// RiskAssessment contains risk assessment information
type RiskAssessment struct {
	Score      float64            `json:"score"` // 0-100
	Level      string             `json:"level"` // low, medium, high, critical
	Violations int64              `json:"total_violations"`
	Details    map[string]float64 `json:"details"` // breakdown by violation type
}

// =============================================================================
// Event Types for Recording (Dashboard-specific, distinct from streaming events)
// =============================================================================

// PolicyViolationRecord represents a policy violation event to be recorded in the database
type PolicyViolationRecord struct {
	ID            string         `json:"id"`
	APIKeyID      string         `json:"api_key_id,omitempty"`
	PolicyID      string         `json:"policy_id"`
	PolicyName    string         `json:"policy_name"`
	ViolationType string         `json:"violation_type"`
	Severity      int            `json:"severity"` // 1=low, 2=medium, 3=high, 4=critical, 5=severe
	Message       string         `json:"message,omitempty"`
	Timestamp     time.Time      `json:"timestamp"`
	Metadata      map[string]any `json:"metadata,omitempty"`
}

// ToolCallRecord represents a tool call event to be recorded in the database
type ToolCallRecord struct {
	ID           string    `json:"id"`
	APIKeyID     string    `json:"api_key_id,omitempty"`
	ToolName     string    `json:"tool_name"`
	Model        string    `json:"model"`
	Provider     string    `json:"provider"`
	Success      bool      `json:"success"`
	ErrorMessage string    `json:"error_message,omitempty"`
	Timestamp    time.Time `json:"timestamp"`
}

// CacheEventRecord represents a cache hit/miss event to be recorded in the database
type CacheEventRecord struct {
	ID           string    `json:"id"`
	APIKeyID     string    `json:"api_key_id,omitempty"`
	Model        string    `json:"model"`
	Hit          bool      `json:"hit"`
	TokensSaved  int64     `json:"tokens_saved"`
	CostSavedUSD float64   `json:"cost_saved_usd"`
	Timestamp    time.Time `json:"timestamp"`
}

// =============================================================================
// Repository Interfaces
// =============================================================================

// AgentDashboardRepository handles agent dashboard data storage and retrieval
type AgentDashboardRepository interface {
	// GetStats retrieves comprehensive dashboard statistics for an agent
	GetStats(ctx context.Context, tenantID, apiKeyID string, startTime, endTime time.Time) (*AgentDashboardStats, error)

	// RecordPolicyViolation records a policy violation event
	RecordPolicyViolation(ctx context.Context, event *PolicyViolationRecord) error

	// RecordToolCall records a tool call event
	RecordToolCall(ctx context.Context, event *ToolCallRecord) error

	// RecordCacheEvent records a cache hit/miss event
	RecordCacheEvent(ctx context.Context, event *CacheEventRecord) error

	// GetPolicyViolations retrieves policy violations for a time range
	GetPolicyViolations(ctx context.Context, tenantID, apiKeyID string, startTime, endTime time.Time) ([]PolicyViolationRecord, error)

	// CleanupOldData removes data older than the retention period (30 days)
	CleanupOldData(ctx context.Context) error
}
